### ISSUE TITLE
slice assignment to instrumented list should check if existing list is passed and copy

### ISSUE THREAD

**Issue by gemerden** [sqlalchemy/sqlalchemy#4990]
while i can easily fix this in my code, this seems a bug, What happens is that this leaves Object.items as an empty list, whatever it was before. It happens in "sqlalchemy\orm\collections.py" on lines 1157-1163:

                if step == 1:
                    for i in range(start, stop, step):
                        if len(self) > start:
                            del self[start]

                    for i, item in enumerate(value):
                        self.insert(i + start, item)

Basically if self is the same InstrumentedList as value, then 'del self[start]' also removes value[start], leaving an empty list to be inserted. This would (unverified) also go wrong with any other slice, as long as self and value are the same object.

This is not the same behavior is in python itself:

L = [1,2,3]
L[:] = L
print(L)

**Comment by zzzeek** [comment]
please work around the issue for now.  the collection replacement are extremely complicated and if you are managing to trip it up, fixing it is typically not trivial, unless you have an interest in working on this.

**Comment by zzzeek** [comment]
also I cannot reproduce.    Please ensure you are on the latest 1.3 version as you may be seeing older behavior.   Below is an MCVE, please modify it to illustrate your failure condition, thanks.

```
from sqlalchemy import *
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import *

Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(String)
    bs = relationship("B")


class B(Base):
    __tablename__ = 'b'
    id = Column(Integer, primary_key=True)
    a_id = Column(ForeignKey("a.id"))
    data = Column(String)


b1, b2, b3 = B(), B(), B()

a1 = A()

a1.bs = bs = [b1, b2, b3]

assert a1.bs == [b1, b2, b3]
a1.bs[:] = bs

assert a1.bs == [b1, b2, b3]

```

**Comment by gemerden** [comment]
the version is 1.3.10 (sry)

I have changed the MCVE so it shows the bug:

from sqlalchemy import *
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import *

Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(String)
    bs = relationship("B")


class B(Base):
    __tablename__ = 'b'
    id = Column(Integer, primary_key=True)
    a_id = Column(ForeignKey("a.id"))
    data = Column(String)


b1, b2, b3 = B(), B(), B()

a1 = A()

a1.bs = [b1, b2, b3]

bs = a1.bs  # changed: assign the InstrumentedList to bs

assert a1.bs == [b1, b2, b3]
assert a1.bs is bs

a1.bs[:] = bs

assert a1.bs == [b1, b2, b3]  # a1.bs == []

In your example bs was just a list not an InstrumentedList.

I think the fix can be:

                if step == 1:
                    value = value[:]  # -> new: copy the value in case "value is self", could also check first
                    for i in range(start, stop, step):
                        if len(self) > start:
                            del self[start]

                    for i, item in enumerate(value):
                        self.insert(i + start, item)

I havent checked

**Comment by zzzeek** [comment]
thank you, that is an easy fix and localized to an infrequently used codepath.

**Comment by zzzeek** [comment]
welp, the tests say otherwise, if "value is self" we should do nothing because otherwise we are generating new history events.

**Comment by zzzeek** [comment]
haven't had breakfast yet so not sure if im missing anything:

```
diff --git a/lib/sqlalchemy/orm/collections.py b/lib/sqlalchemy/orm/collections.py
index 4b096c152..5008f5727 100644
--- a/lib/sqlalchemy/orm/collections.py
+++ b/lib/sqlalchemy/orm/collections.py
@@ -1201,6 +1201,8 @@ def _list_decorators():
                     stop += len(self)
 
                 if step == 1:
+                    if value is self:
+                        return
                     for i in range(start, stop, step):
                         if len(self) > start:
                             del self[start]

```

**Comment by zzzeek** [comment]
this whole case is pretty broken because for example we don't support things like a.bs[0:1] = bs, as this kind of thing implies duplicate entries

**Comment by zzzeek** [comment]
but will do the full replace case for now, at least

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **master** branch:

**Skip on slice assignment to self** https://gerrit.sqlalchemy.org/1580

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **rel_1_3** branch:

**Skip on slice assignment to self** https://gerrit.sqlalchemy.org/1581

**Comment by gemerden** [comment]
> this whole case is pretty broken because for example we don't support things like a.bs[0:1] = bs, as this kind of thing implies duplicate entries

This currently sends the copy mechanism in an infinite loop ... ;-), and adding "value = value[:]" would create duplicates.

just returning would act like normal python in the [:] case and like a copy and deduplicate in the [a:b] case, as far as i can see, so ... great! ... :-)

**Comment by zzzeek** [comment]
are you suggesting a further change?   i didnt spend much time trying to understand the [a:b] case as well as our test suite which someone else wrote a long time ago and is also not that clear to me

**Comment by gemerden** [comment]
Nope, i was agreeing, sorry about the misunderstanding, the:

    If value is self:
        return

no-op solution would also cover the [i:j] case (i shouldn’t have used a and b ;-) .

My point was that since a no-op would leave the list intact, and the [i:j] can only create duplicates, this is the closest to ‘normal’ python behavior one could get.

The

If value is self:
        value = value[:]

solution would create duplicates however.

**Comment by zzzeek** [comment]
yes that was what the test suite found.

**Comment by zzzeek** [comment]
hey guess what, doesnt work at all this way in python 2.    that codepath doesn't even happen.

**Comment by zzzeek** [comment]
or something.  not sure

**Comment by zzzeek** [comment]
yep, it's not called.

**Comment by zzzeek** [comment]
oh it calls `__setslice__`.
